File: Java/src/storm/xmlbinder/binder/ObjectBinder.java
/*
 * Copyright Julien Mialon, Millian Poquet, Soumia Betta, Tarik Abas.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package storm.xmlbinder.binder;

import org.w3c.dom.Document;
import org.w3c.dom.Node;
import storm.xmlbinder.XmlBinderException;

/**
 * Binder to bind xml element to a new Object.
 *
 * @author Storm <j-storm@hotmail.fr>
 */
public class ObjectBinder extends AbstractBinder
{
	/**
	 * The name of the class to create the object.
	 */
	protected String m_className = "";

	/**
	 * The content binder to use if needed.
	 */
	protected ContentBinder m_contentBinder = null;

	/**
	 * Construct an instance without a content binder.
	 * Assuming that the xml element is not a text node and has child element.
	 *
	 * @param _fieldName : the name of the field.
	 * @param _className : the name of the class to create object.
	 */
	public ObjectBinder(String _fieldName, String _className)
	{
		super(_fieldName);
		this.m_className = _className;
	}

	/**
	 * Construct an instance with a content binder.
	 * Assuming that the xml element is a text node.
	 *
	 * @param _fieldName     : the name of the field.
	 * @param _className     : the name of the class to create object.
	 * @param _contentBinder : the content binder to use.
	 */
	public ObjectBinder(String _fieldName, String _className, ContentBinder _contentBinder)
	{
		this(_fieldName, _className);
		this.m_contentBinder = _contentBinder;
	}

	/*
	 * (non-Javadoc)
	 * @see storm.xmlbinder.binder.AbstractBinder#read(java.lang.Object, java.lang.String)
	 */
	@Override
	public Object read(Object _data, String _content) throws XmlBinderException
	{
		Object newData = createObject(this.m_className);
		setField(_data, this.m_fieldName, newData);

		if (this.m_contentBinder != null)
		{
			this.m_contentBinder.read(newData, _content);
		}

		return newData;
	}

	/*
	 * (non-Javadoc)
	 * @see storm.xmlbinder.binder.AbstractBinder#write(java.lang.Object, org.w3c.dom.Node, org.w3c.dom.Document)
	 */
	@Override
	public void write(Object _data, Node _node, Document _document) throws XmlBinderException
	{
		//Xml element already created, just have to map the content binder if needed
		if (this.m_contentBinder != null)
		{
			this.m_contentBinder.write(_data, _node, _document);
		}
	}

}

File: C#/Binder/ObjectBinder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using XmlBinder.Interface;
using XmlBinder.Tools;

namespace XmlBinder.Binder
{
	/// <summary>
	/// Binder used to create an object of the specified type when the xml tag is encoutered.
	/// </summary>
	public class ObjectBinder : XmlBinder.Abstract.AbstractBinder
	{
		/// <summary>
		/// Type of the object to create.
		/// </summary>
		protected Type m_classType;

		/// <summary>
		/// Binder to bind content of the xml tag.
		/// </summary>
		protected BinderInterface m_contentBinder;

		/// <summary>
		/// Type to create object.
		/// </summary>
		public Type ClassType
		{
			get
			{
				return m_classType;
			}
		}

		/// <summary>
		/// Binder to bind content of xml tag if needed.
		/// </summary>
		public BinderInterface ContentBinder
		{
			get
			{
				return m_contentBinder;
			}
		}

		/// <summary>
		/// Build a binder with specified class type and content binder.
		/// </summary>
		/// <param name="_fieldName">The field to bind the new object to.</param>
		/// <param name="_classType">The type of the new object.</param>
		/// <param name="_contentBinder">A Binder to bind the content of xml tag if needed.</param>
		public ObjectBinder(string _fieldName, Type _classType, BinderInterface _contentBinder = null) : base(_fieldName)
		{
			m_classType = _classType;
			m_contentBinder = _contentBinder;
		}

		/// <see cref="BinderInterface.read()"/>
		public override object read(object _source, string _content)
		{
			object data = Activator.CreateInstance(ClassType);
			FieldAccess.setField(_source, FieldName, data);
			if (ContentBinder != null)
			{
				ContentBinder.read(data, _content);
			}
			return data;
		}

		/// <see cref="BinderInterface.write()"/>
		public override string write(object _source)
		{
			if (ContentBinder != null)
			{
				object data = ContentBinder.getAffectedField(_source);
				//object data = FieldAccess.getField(_source, FieldName);
				return ContentBinder.write(data);
			}
			return "";
		}
	}
}
